Runtime/Behaviours/CinemachineConfiner2D.cs

```diff
@@ -0,0 +1,484 @@
+                    displacement -= extra.m_DampedDisplacement;
+                }
+                state.PositionCorrection += displacement;
+            }
+        }
+
+        /// <summary>
+        /// Calculates half frustum height for orthographic or perspective camera.
+        /// For more info on frustum height, see <see cref="docs.unity3d.com/Manual/FrustumSizeAtDistance.html"/> 
+        /// </summary>
+        /// <param name="state">CameraState for checking if Orthographic or Perspective</param>
+        /// <param name="vcam">vcam, to check its position</param>
+        /// <returns>Frustum height of the camera</returns>
+        private float CalculateHalfFrustumHeight(in CameraState state, in float cameraPosLocalZ)
+        {
+            float frustumHeight;
+            if (state.Lens.Orthographic)
+            {
+                frustumHeight = state.Lens.OrthographicSize;
+            }
+            else
+            {
+                // distance between the collider's plane and the camera
+                float distance = cameraPosLocalZ;
+                frustumHeight = distance * Mathf.Tan(state.Lens.FieldOfView * 0.5f * Mathf.Deg2Rad);
+            }
+
+            return Mathf.Abs(frustumHeight);
+        }
+        
+        /// <summary>
+        /// Confines input 2D point within the confined area.
+        /// </summary>
+        /// <param name="positionToConfine">2D point to confine</param>
+        /// <returns>Confined position</returns>
+        private Vector2 ConfinePoint(Vector2 positionToConfine, in List<List<Vector2>> pathCache,
+            in bool hasBone, in float windowWidth, in float windowHeight)
+        {
+            if (ShrinkablePolygon.IsInside(pathCache, positionToConfine))
+            {
+                return positionToConfine;
+            }
+            bool outsideOfOriginal = !ShrinkablePolygon.IsInside(m_shapeCache.m_OriginalPath, positionToConfine);
+
+            Vector2 closest = positionToConfine;
+            float minDistance = float.MaxValue;
+            for (int i = 0; i < pathCache.Count; ++i)
+            {
+                int numPoints = pathCache[i].Count;
+                for (int j = 0; j < numPoints; ++j)
+                {
+                    Vector2 v0 = pathCache[i][j];
+                    Vector2 v = pathCache[i][(j + 1) % numPoints];
+                    Vector2 c = Vector2.Lerp(v0, v, positionToConfine.ClosestPointOnSegment(v0, v));
+                    Vector2 difference = positionToConfine - c;
+                    float distance = Vector2.SqrMagnitude(difference);
+                    if (Mathf.Abs(difference.x) > windowWidth || Mathf.Abs(difference.y) > windowHeight)
+                    {
+                        // penalty for points from which the target is not visible, prefering visibility over proximity
+                        distance += m_confinerBaker.m_sqrPolygonDiagonal; 
```

Comment:
You're going to have to explain this to me.  I don't understand why you give it such a large penalty.  What if it really is the best point?

Reply:
Best points are points from which the target is visible in the camera window.
If the target is not visible from the point, then I give it a penalty.

This essentially sorts my points into two groups:
- Group 1: points from which target is visible
- Group 2: points from which target is not visible

If Group 1 is not empty, then I will choose the closest point in Group 1.
If Group 1 is empty, then I will choose the closest point in Group 2.